setuptools/build_meta.py

```diff
@@ -110,51 +95,88 @@ def _file_with_extension(directory, extension):
     return file
 
 
-def _open_setup_script(setup_script):
-    if not os.path.exists(setup_script):
-        # Supply a default setup.py
-        return io.StringIO(u"from setuptools import setup; setup()")
-
-    return getattr(tokenize, 'open', open)(setup_script)
-
-
 @contextlib.contextmanager
 def suppress_known_deprecation():
     with warnings.catch_warnings():
         warnings.filterwarnings('ignore', 'setup.py install is deprecated')
         yield
 
 
+@contextlib.contextmanager
+def _patch_distutils_core():
+    """Make sure distutils.core uses the latest enhancements"""
+    orig_exec = exec
+    if hasattr(distutils.core, "run_commands"):
+        yield  # do nothing, already using the improved version of distutils
+        return
+
+    def _exec(code, global_vars):
+        try:
+            fid, tmp = tempfile.mkstemp(suffix=f"{uuid4()}-setup.py", text=False)
+            os.close(fid)  # Ignore the low level API
+            with open(tmp, "wb") as f:
+                f.write(code)
+            with tokenize.open(tmp) as f:
+                code = f.read().replace(r'\r\n', r'\n')
+        finally:
+            os.remove(tmp)
```

Comment:
The implementation is quite straightforward here https://github.com/python/cpython/blob/81c7204/Lib/tokenize.py#L392-L405. and could be emulated a bit more efficiently with smth like
```suggestion
        from io import BytesIO, TextIOWrapper

        with BytesIO(code) as bytes_code_buffer:
            encoding, _lines = detect_encoding(bytes_code_buffer.readline)
            bytes_code_buffer.seek(0)
            with TextIOWrapper(buffer, encoding, line_buffering=True) as text:
                code = text.read().replace(r'\r\n', r'\n')
```

Reply:
Thank you very much for the improvement!